### PR TITLE
py/grammar.h: Add matmul operator (@) as per PEP 465.

### PR DESCRIPTION
**Pull request by jimmo**
Maps to `__matmul__`, not supported for any built-in type (but intended for `unumpy.ndarray`).
Same precedence as `*` and left associative.
Also adds corresponding `__imatmul__` and `__rmatmul__`.

This is +80 bytes on ports/minimal. Also a breaking change to the bytecode as the op table changes (but my understanding is that this is OK between micropython versions?)

### CODE REVIEW

**Comment by jimmo** [comment]
+20 bytes according to the Travis results.

Appveyor failed due to CPython 3.4 on the Windows build. Will make the test conditional.

**Comment by dpgeorge** [comment]
> Appveyor failed due to CPython 3.4 on the Windows build. Will make the test conditional.

For such tests a .py.exp file is usually provided (see eg `tests/basics/python36.py`)

**Comment by jimmo** [comment]
Thanks :) Done

**Comment by dpgeorge** [comment]
This is the code size change I measure locally:
```
   bare-arm:   +16 +0.024% 
minimal x86:   +80 +0.052% 
   unix x64:  +176 +0.036%
unix nanbox:  +144 +0.033% 
      stm32:   +32 +0.009% 
     cc3200:   +24 +0.013% 
    esp8266:   +80 +0.012% 
      esp32:   +36 +0.003%
```
So it might be worth thinking of ways to minimise this impact.  It could be made compile-time configurable but that's going to be rather messy.

**Comment by dpgeorge** [comment]
> Also a breaking change to the bytecode as the op table changes (but my understanding is that this is OK between micropython versions?)

Bytecode version already changed since the last release, so there is scope to make a change again before the next release.

File: py/compile.c [hunk 1]
```diff
@@ -2006,6 +2006,7 @@ STATIC void compile_expr_stmt(compiler_t *comp, mp_parse_node_struct_t *pns) {
                 case MP_TOKEN_DEL_DBL_SLASH_EQUAL: op = MP_BINARY_OP_INPLACE_FLOOR_DIVIDE; break;
                 case MP_TOKEN_DEL_SLASH_EQUAL: op = MP_BINARY_OP_INPLACE_TRUE_DIVIDE; break;
                 case MP_TOKEN_DEL_PERCENT_EQUAL: op = MP_BINARY_OP_INPLACE_MODULO; break;
+                case MP_TOKEN_DEL_AT_EQUAL: op = MP_BINARY_OP_INPLACE_MATMUL; break;
```
Comment (dpgeorge) [inline py/compile.c]: It might be possible to compress this switch statement into a small array with a single byte lookup (if so it should be a separate PR to fully evaluate it).

File: py/compile.c [hunk 2]
```diff
@@ -2212,6 +2213,7 @@ STATIC void compile_term(compiler_t *comp, mp_parse_node_struct_t *pns) {
             case MP_TOKEN_OP_DBL_SLASH: op = MP_BINARY_OP_FLOOR_DIVIDE; break;
             case MP_TOKEN_OP_SLASH:     op = MP_BINARY_OP_TRUE_DIVIDE; break;
             case MP_TOKEN_OP_PERCENT:   op = MP_BINARY_OP_MODULO; break;
+            case MP_TOKEN_OP_AT:        op = MP_BINARY_OP_MATMUL; break;
             case MP_TOKEN_OP_DBL_LESS:  op = MP_BINARY_OP_LSHIFT; break;
```
Comment (dpgeorge) [inline py/compile.c]: This switch might also be compressed.

**Comment by pfalcon** [comment]
> So it might be worth thinking of ways to minimise this impact. It could be made compile-time configurable but that's going to be rather messy.

+1, this is a pet operator of numpy, which I never saw to be used anywhere else. (Then maybe numpy support for MicroPython would be important, but I didn't see someone arguing that. And then support for @ would be again among the last things to care about, after 95% of other functionality.)

**Comment by stinos** [comment]
> numpy support for MicroPython would be important, but I didn't see someone arguing that.

Well, it would be pretty awesome. We actually have more than one usecase here where we have uPy code dump data to a binary format, then launch a CPython process to deal with it in numpy and get results from that back into uPy. However that's just us, and numpy is like the opposite of 'micro', so I can see why it's not there yet.

**Comment by jimmo** [comment]
@dpgeorge Good thinking on the switch statements. Especially the non-inplace one benefitted because it could take advantage of the table already used in parse.c.

```
   bare-arm:    -4 -0.006% 
minimal x86:   +48 +0.031% 
   unix x64:  +128 +0.026% [incl +32(data)]
```

@pfalcon Yes it's highly specialised for numpy, it's literally called "matmul" for a reason. Actually it's pretty high on the list of things to care about because it's a good feature that people use, it means that existing code will work without modification, and it avoids an error-prone conversion of `a @ b` to `numpy.matmul(a, b)`.

**Comment by dpgeorge** [comment]
See #4947 for a reworked version of this.

**Comment by dpgeorge** [comment]
Superseded by #4947
